rancher2_app_v2 throws error about inconsistent results

When trying to add a new `rancher2_app_v2` resource for Dynatrace, I got the following error on apply:

```
04-Dec-2020 14:44:17 | Error: Provider produced inconsistent result after apply
-- | --
04-Dec-2020 14:44:17 |  
04-Dec-2020 14:44:17 | When applying changes to
04-Dec-2020 14:44:17 | module.rancher_cluster["glent-dev"].module.rancher_cluster_configuration[0].rancher2_app_v2.dynatrace[0],
04-Dec-2020 14:44:17 | provider "registry.terraform.io/rancher/rancher2" produced an unexpected new
04-Dec-2020 14:44:17 | value: Root resource was present, but now absent.
04-Dec-2020 14:44:17 |  
04-Dec-2020 14:44:17 | This is a bug in the provider, which should be reported in the provider's own
04-Dec-2020 14:44:17 | issue tracker.
04-Dec-2020 14:44:17 |  
04-Dec-2020 14:44:17 | make: *** [apply-plan] Error 1
```

I got this by running the following plan:

```
04-Dec-2020 14:43:20 | Terraform will perform the following actions:
-- | --
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | # module.rancher_cluster["glent-dev"].module.rancher_cluster_configuration[0].rancher2_app_v2.dynatrace[0] will be created
04-Dec-2020 14:43:20 | + resource "rancher2_app_v2" "dynatrace" {
04-Dec-2020 14:43:20 | + annotations                 = (known after apply)
04-Dec-2020 14:43:20 | + chart_name                  = "dynatrace-oneagent-operator"
04-Dec-2020 14:43:20 | + chart_version               = "0.8.000"
04-Dec-2020 14:43:20 | + cleanup_on_fail             = false
04-Dec-2020 14:43:20 | + cluster_id                  = "c-krmhh"
04-Dec-2020 14:43:20 | + cluster_name                = (known after apply)
04-Dec-2020 14:43:20 | + disable_hooks               = false
04-Dec-2020 14:43:20 | + disable_open_api_validation = false
04-Dec-2020 14:43:20 | + force_upgrade               = false
04-Dec-2020 14:43:20 | + id                          = (known after apply)
04-Dec-2020 14:43:20 | + labels                      = (known after apply)
04-Dec-2020 14:43:20 | + name                        = "dynatrace"
04-Dec-2020 14:43:20 | + namespace                   = (known after apply)
04-Dec-2020 14:43:20 | + project_id                  = "c-krmhh:p-bhd7q"
04-Dec-2020 14:43:20 | + repo_name                   = "rancher-partner-charts"
04-Dec-2020 14:43:20 | + values                      = <<~EOT
04-Dec-2020 14:43:20 | ---
04-Dec-2020 14:43:20 | oneagent:
04-Dec-2020 14:43:20 | apiUrl: "REDACTED"
04-Dec-2020 14:43:20 | secret:
04-Dec-2020 14:43:20 | apiToken: "REDACTED"
04-Dec-2020 14:43:20 | paasToken: "REDACTED"
04-Dec-2020 14:43:20 | EOT
04-Dec-2020 14:43:20 | + wait                        = false
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | # module.rancher_cluster["glent-dev"].module.rancher_cluster_configuration[0].rancher2_namespace.dynatrace will be created
04-Dec-2020 14:43:20 | + resource "rancher2_namespace" "dynatrace" {
04-Dec-2020 14:43:20 | + annotations      = (known after apply)
04-Dec-2020 14:43:20 | + id               = (known after apply)
04-Dec-2020 14:43:20 | + labels           = (known after apply)
04-Dec-2020 14:43:20 | + name             = "dynatrace"
04-Dec-2020 14:43:20 | + project_id       = "c-krmhh:p-bhd7q"
04-Dec-2020 14:43:20 | + wait_for_cluster = false
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | + resource_quota {
04-Dec-2020 14:43:20 | + limit {
04-Dec-2020 14:43:20 | + config_maps              = (known after apply)
04-Dec-2020 14:43:20 | + limits_cpu               = (known after apply)
04-Dec-2020 14:43:20 | + limits_memory            = (known after apply)
04-Dec-2020 14:43:20 | + persistent_volume_claims = (known after apply)
04-Dec-2020 14:43:20 | + pods                     = (known after apply)
04-Dec-2020 14:43:20 | + replication_controllers  = (known after apply)
04-Dec-2020 14:43:20 | + requests_cpu             = (known after apply)
04-Dec-2020 14:43:20 | + requests_memory          = (known after apply)
04-Dec-2020 14:43:20 | + requests_storage         = (known after apply)
04-Dec-2020 14:43:20 | + secrets                  = (known after apply)
04-Dec-2020 14:43:20 | + services                 = (known after apply)
04-Dec-2020 14:43:20 | + services_load_balancers  = (known after apply)
04-Dec-2020 14:43:20 | + services_node_ports      = (known after apply)
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | Plan: 2 to add, 0 to change, 0 to destroy.
04-Dec-2020 14:43:20 |  
```

The plan was the result of the following code:

```
resource "rancher2_namespace" "dynatrace" {
  name       = "dynatrace"
  project_id = module.plumbing_workspace.project_id
}

resource "rancher2_app_v2" "dynatrace" {
  count = var.dynatrace_enable ? 1 : 0

  name       = "dynatrace"
  cluster_id = var.cluster_id

  repo_name     = "rancher-partner-charts"
  chart_name    = "dynatrace-oneagent-operator"
  chart_version = "0.8.000"

  project_id = module.plumbing_workspace.project_id
  namespace  = rancher2_namespace.dynatrace.id

  values = templatefile(
    "${path.module}/dynatrace/config.yml.tpl",
    var.dynatrace_settings
  )
}
```

The app was created successfully, although not all resources were rolled out successfully. When rerunning the plan/apply, it tries to create a new app, even though the old one is still there.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,15 +5,23 @@ FEATURES:
 * **New Argument:** `rancher2_node_template.hetzner_config` - (Optional) Hetzner config for the Node Template (list maxitems:1)
 * **New Argument:** `rancher2_cluster.rke_config.dns.linear_autoscaler_params` - (Optional) LinearAutoScalerParams dns config (list Maxitem: 1)
 * **New Argument:** `rancher2_cluster.rke_config.dns.update_strategy` - (Optional) DNS update strategy (list Maxitems: 1)
+* **New Argument:** `rancher2_notifier.dingtalk_config` - (Optional) Dingtalk config for notifier (list maxitems:1)
+* **New Argument:** `rancher2_notifier.msteams_config` - (Optional) MSTeams config for notifier (list maxitems:1)
 
 ENHANCEMENTS:
 
-
+* Updated `rancher2_app_v2.chart_version` as optional/computed argument. Deploying latest app v2 version if `chart_version` is not provided
+* Updated `rancher2_app_v2.wait` default value to `true`
 
 BUG FIXES:
 
 * Fixed `rancher2_cluster` resource, added retry when enabling cluster monitoring and got apierr 500. https://github.com/rancher/rancher/issues/30188
 * Fixed `rancher2_cluster` datasource error, when `rke_config.services.kube_api.secrets_encryption_config.custom_config` or `rke_config.services.kube_api.event_rate_limit.configuration` are set. https://github.com/rancher/terraform-provider-rancher2/issues/546
+* Fixed `rancher2_cluster_template` required argument definition on docs
+* Fixed `Apps & marketplace` guide for Rancher v2.5.0 format
+* Fixed doc examples for activedirectory, freeipa and openldap auth providers
+* Fixed `rancher2_app_v2` resource to properly pass global values to sub charts. https://github.com/rancher/terraform-provider-rancher2/issues/545
+* Fixed `rancher2_app_v2` resource to don't override name nor namespace on App v2 not certified by rancher
 
 ## 1.10.6 (November 11, 2020)
 

--- a/docs/guides/apps_marketplace.md
+++ b/docs/guides/apps_marketplace.md
@@ -101,7 +101,7 @@ This resource has the following arguments definition:
 * `disable_hooks` - (Optional) Disable app v2 chart hooks. Default: `false` (bool)
 * `disable_open_api_validation` - (Optional) Disable app V2 Open API Validation. Default: `false` (bool)
 * `force_upgrade` - (Optional) Force app V2 chart upgrade. Default: `false` (bool)
-* `wait` - (Optional) Wait until app is deployed. Default: `false` (bool)
+* `wait` - (Optional) Wait until app is deployed. Default: `true` (bool)
 * `annotations` - (Optional/Computed) Annotations for the app v2 (map)
 * `labels` - (Optional/Computed) Labels for the app v2 (map)
 
@@ -199,7 +199,6 @@ resource "rancher2_app_v2" "rancher-cis-benchmark" {
   repo_name = "rancher-charts"
   chart_name = "rancher-cis-benchmark"
   chart_version = "1.0.100"
-  wait = true
 }
 ```
 
@@ -230,7 +229,6 @@ s3:
   insecureTLSSkipVerify: false
   region: ""
 EOF
-  wait = true
 }
 ```
 
@@ -244,7 +242,19 @@ resource "rancher2_app_v2" "rancher-logging" {
   repo_name = "rancher-charts"
   chart_name = "rancher-logging"
   chart_version = "3.6.000"
-  wait = true
+}
+```
+
+* Deploy Rancher Longhorn
+
+```hcl
+resource "rancher2_app_v2" "longhorn" {  
+  cluster_id = "<CLUSTER_ID>"
+  name = "rancher-longhorn"
+  namespace = "longhorn-system"
+  repo_name = "rancher-charts"
+  chart_name = "longhorn"
+  chart_version = "1.0.201"
 }
 ```
 

--- a/docs/guides/apps_marketplace.md
+++ b/docs/guides/apps_marketplace.md
@@ -45,10 +45,10 @@ These are some examples how to use a new catalog v2:
 
 ```hcl
 resource "rancher2_catalog_v2" "foo" {
-  cluster_id = "&lt;CLUSTER_ID&gt;""
+  cluster_id = "<CLUSTER_ID>"
   name = "foo"
-  git_repo = "&lt;GIT_REPO_URL&gt;"
-  git_branch = "&lt;GIT_BRANCH&gt;"
+  git_repo = "<GIT_REPO_URL>"
+  git_branch = "<GIT_BRANCH>"
 }
 ```
 
@@ -56,9 +56,9 @@ resource "rancher2_catalog_v2" "foo" {
 
 ```hcl
 resource "rancher2_catalog_v2" "foo-url" {
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "foo-url"
-  url = "https://&lt;CATALOG_URL&gt;"
+  url = "https://<CATALOG_URL>"
 }
 ```
 
@@ -67,7 +67,7 @@ resource "rancher2_catalog_v2" "foo-url" {
 ```hcl
 ## Define catalog at tf file
 resource "rancher2_catalog_v2" "rancher-charts" {
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-charts"
   git_repo = "https://git.rancher.io/charts"
   git_branch = "main"
@@ -80,7 +80,7 @@ resource "rancher2_catalog_v2" "rancher-charts" {
 
 ```hcl
 data "rancher2_catalog_v2" "foo" {
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "foo"
 }
 ```
@@ -107,19 +107,23 @@ This resource has the following arguments definition:
 
 ### Examples
 
-These are some examples how to deploy some Rancher cerfified apps v2:
+These are some examples how to deploy some Rancher cerfified Apps v2:
 
-* Deploy Rancher monitoring
+**Note** Apps v2 certified by Rancher should be installed within fixed app name and namespace defined by chart annotations, `catalog.cattle.io/release-name` and `catalog.cattle.io/namespace`. The provider will be forcing them on Rancher cerfified Apps v2 installation, but generating a diff on next `terraform apply`, if app v2 name and/or namespace is customized by user. Adjusting the values at tf file should supress the diff
+
+**Note** Latest app v2 version will be deployed, if `chart_version` is not provided
+
+* `rancher-monitoring` - Deploy Rancher monitoring
 
 ```hcl
 resource "rancher2_app_v2" "rancher-monitoring" {
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-monitoring"
   namespace = "cattle-monitoring-system"
   repo_name = "rancher-charts"
   chart_name = "rancher-monitoring"
   chart_version = "9.4.200"
-  values = &lt;&lt;EOF
+  values = <<EOF
 prometheus:
   prometheusSpec:
     requests:
@@ -149,17 +153,17 @@ EOF
 #      memory: "250Mi"
 ```
 
-* Deploy Rancher istio and Rancher monitoring as requirement
+* `rancher-istio` - Deploy Rancher istio and Rancher monitoring as requirement
 
 ```hcl
 resource "rancher2_app_v2" "rancher-monitoring" {
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-monitoring"
   namespace = "cattle-monitoring-system"
   repo_name = "rancher-charts"
   chart_name = "rancher-monitoring"
   chart_version = "9.4.200"
-  values = &lt;&lt;EOF
+  values = <<EOF
 alertmanager:
   alertmanagerSpec:
     enabled: false
@@ -176,7 +180,7 @@ EOF
 resource "rancher2_app_v2" "rancher-istio" {
   depends_on = [rancher2_app_v2.rancher-monitoring] # Rancher-istio requires rancher-monitoring
   
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-istio"
   namespace = "istio-system"
   repo_name = "rancher-charts"
@@ -185,13 +189,13 @@ resource "rancher2_app_v2" "rancher-istio" {
 }
 ```
 
-* Deploy Rancher cis benchmark
+* `rancher-cis-benchmark` - Deploy Rancher cis benchmark
 
 ```hcl
 resource "rancher2_app_v2" "rancher-cis-benchmark" {  
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-cis-benchmark"
-  namespace = "istio-system"
+  namespace = "cis-operator-system"
   repo_name = "rancher-charts"
   chart_name = "rancher-cis-benchmark"
   chart_version = "1.0.100"
@@ -199,17 +203,17 @@ resource "rancher2_app_v2" "rancher-cis-benchmark" {
 }
 ```
 
-* Deploy Rancher backup
+* `rancher-backup` - Deploy Rancher backup
 
 ```hcl
 resource "rancher2_app_v2" "rancher-backup" {  
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-backup"
   namespace = "cattle-resources-system"
   repo_name = "rancher-charts"
   chart_name = "rancher-backup"
   chart_version = "1.0.200"
-  values = &lt;&lt;EOF
+  values = <<EOF
 persistence:
   enabled: false
   size: 2Gi
@@ -230,16 +234,44 @@ EOF
 }
 ```
 
-* Deploy Rancher logging
+* `rancher-logging` - Deploy Rancher logging
 
 ```hcl
 resource "rancher2_app_v2" "rancher-logging" {  
-  cluster_id = "&lt;CLUSTER_ID&gt;"
+  cluster_id = "<CLUSTER_ID>"
   name = "rancher-logging"
   namespace = "cattle-logging-system"
   repo_name = "rancher-charts"
   chart_name = "rancher-logging"
   chart_version = "3.6.000"
+  wait = true
+}
+```
+
+* `longhorn` - Deploy Rancher Longhorn
+
+```hcl
+resource "rancher2_app_v2" "longhorn" {  
+  cluster_id = "<CLUSTER_ID>"
+  name = "rancher-longhorn"
+  namespace = "longhorn-system"
+  repo_name = "rancher-charts"
+  chart_name = "longhorn"
+  chart_version = "1.0.201"
+  wait = true
+}
+```
+
+* `rancher-gatekeeper` - Deploy OPA Gatekeeper
+
+```hcl
+resource "rancher2_app_v2" "rancher-gatekeeper" {  
+  cluster_id = "<CLUSTER_ID>"
+  name = "rancher-gatekeeper"
+  namespace = "cattle-gatekeeper-system"
+  repo_name = "rancher-charts"
+  chart_name = "rancher-gatekeeper"
+  chart_version = "3.1.101"
   wait = true
 }
 ```

--- a/docs/resources/app_v2.md
+++ b/docs/resources/app_v2.md
@@ -30,7 +30,7 @@ The following arguments are supported:
 * `namespace` - (Required/ForceNew) The namespace of the app v2 (string)
 * `repo_name` - (Required) Repo name (string)
 * `chart_name` - (Required) The app v2 chart name (string)
-* `chart_version` - (Optional) The app v2 chart version (string)
+* `chart_version` - (Optional/Computed) The app v2 chart version (string)
 * `project_id` - (Optional) Deploy the app v2 within project ID (string)
 * `values` - (Optional/Sensitive) The app v2 values yaml. Yaml format is required (string)
 * `cleanup_on_fail` - (Optional) Cleanup app v2 on failed chart upgrade. Default: `false` (bool)

--- a/docs/resources/app_v2.md
+++ b/docs/resources/app_v2.md
@@ -37,7 +37,7 @@ The following arguments are supported:
 * `disable_hooks` - (Optional) Disable app v2 chart hooks. Default: `false` (bool)
 * `disable_open_api_validation` - (Optional) Disable app V2 Open API Validation. Default: `false` (bool)
 * `force_upgrade` - (Optional) Force app V2 chart upgrade. Default: `false` (bool)
-* `wait` - (Optional) Wait until app is deployed. Default: `false` (bool)
+* `wait` - (Optional) Wait until app is deployed. Default: `true` (bool)
 * `annotations` - (Optional/Computed) Annotations for the app v2 (map)
 * `labels` - (Optional/Computed) Labels for the app v2 (map)
 

--- a/docs/resources/auth_config_activedirectory.md
+++ b/docs/resources/auth_config_activedirectory.md
@@ -18,6 +18,8 @@ resource "rancher2_auth_config_activedirectory" "activedirectory" {
   service_account_password = "<SERVICE_PASSWORD>"
   user_search_base = "<SEARCH_BASE>"
   port = <ACTIVEDIRECTORY_PORT>
+  test_username = "<USER_NAME>"
+  test_password = "<USER_PASSWORD>"
 }
 ```
 

--- a/docs/resources/auth_config_freeipa.md
+++ b/docs/resources/auth_config_freeipa.md
@@ -18,6 +18,8 @@ resource "rancher2_auth_config_freeipa" "freeipa" {
   service_account_password = "<SERVICE_PASSWORD>"
   user_search_base = "<SEARCH_BASE>"
   port = <FREEIPA_PORT>
+  test_username = "<USER_NAME>"
+  test_password = "<USER_PASSWORD>"
 }
 ```
 

--- a/docs/resources/auth_config_openldap.md
+++ b/docs/resources/auth_config_openldap.md
@@ -18,6 +18,8 @@ resource "rancher2_auth_config_openldap" "openldap" {
   service_account_password = "<SERVICE_PASSWORD>"
   user_search_base = "<SEARCH_BASE>"
   port = <OPENLDAP_PORT>
+  test_username = "<USER_NAME>"
+  test_password = "<USER_PASSWORD>"
 }
 ```
 

--- a/docs/resources/cluster_alert_group.md
+++ b/docs/resources/cluster_alert_group.md
@@ -52,7 +52,7 @@ The following attributes are exported:
 
 #### Attributes
 
-* `notifier_type` - (Computed) Recipient notifier ID. Supported values : `"pagerduty" | "slack" | "email" | "webhook" | "wechat"` (string)
+* `notifier_type` - (Computed) Recipient notifier ID. Supported values : `"dingtalk" | "msteams" | "pagerduty" | "slack" | "email" | "webhook" | "wechat"` (string)
 
 ## Timeouts
 

--- a/docs/resources/cluster_template.md
+++ b/docs/resources/cluster_template.md
@@ -148,7 +148,7 @@ resource "rancher2_cluster_template" "foo" {
 #### Arguments
 
 * `name` - (Required) The cluster template revision name (string)
-* `cluster_config` - (Optional) Cluster configuration (list maxitem: 1)
+* `cluster_config` - (Required) Cluster configuration (list maxitem: 1)
 * `default` - (Optional) Default cluster template revision. Default `false` (bool)
 * `enabled` - (Optional) Enable cluster template revision. Default `true` (bool)
 * `questions` - (Optional) Cluster template questions (list)
@@ -173,7 +173,7 @@ resource "rancher2_cluster_template" "foo" {
 * `enable_cluster_alerting` - (Optional) Enable built-in cluster alerting. Default: `false` (bool)
 * `enable_cluster_monitoring` - (Optional) Enable built-in cluster monitoring. Default: `false` (bool)
 * `enable_network_policy` - (Optional) Enable project network isolation. Default: `false` (bool)
-* `rke_config` - (Optional/Computed) Rancher Kubernetes Engine Config (list maxitems: 1)
+* `rke_config` - (Required) Rancher Kubernetes Engine Config (list maxitems: 1)
 * `scheduled_cluster_scan`- (Optional) Cluster scheduled cis scan. For Rancher v2.4.0 or above (List MaxItem:1)
 * `windows_prefered_cluster` - (Optional) Windows prefered cluster. Default: `false` (bool)
 

--- a/rancher2/config.go
+++ b/rancher2/config.go
@@ -800,8 +800,8 @@ func (c *Config) DeleteAppV2(clusterID string, app *AppV2) error {
 }
 
 func (c *Config) InfoAppV2(clusterID, repoName, chartName, chartVersion string) (*ClusterRepo, *types2.ChartInfo, error) {
-	if repoName == "" || chartName == "" || chartVersion == "" {
-		return nil, nil, fmt.Errorf("Catalog V2 id, chart name and chart version should be provided")
+	if repoName == "" || chartName == "" {
+		return nil, nil, fmt.Errorf("Catalog V2 id and chart name should be provided")
 	}
 	repo, err := c.GetCatalogV2ByID(clusterID, repoName)
 	if err != nil {
@@ -817,7 +817,10 @@ func (c *Config) InfoAppV2(clusterID, repoName, chartName, chartVersion string) 
 	if resource.Links == nil && len(resource.Links[link]) == 0 {
 		return nil, nil, fmt.Errorf("failed to get chart info %s:%s from catalog v2 %s", chartName, chartVersion, repoName)
 	}
-	resource.Links[link] = resource.Links[link] + "&chartName=" + chartName + "&version=" + chartVersion
+	resource.Links[link] = resource.Links[link] + "&chartName=" + chartName
+	if len(chartVersion) > 0 {
+		resource.Links[link] = resource.Links[link] + "&version=" + chartVersion
+	}
 
 	client, err := c.CatalogV2Client(clusterID)
 	if err != nil {

--- a/rancher2/config.go
+++ b/rancher2/config.go
@@ -1780,7 +1780,11 @@ func (c *Config) GetRecipientByNotifier(id string) (*managementClient.Recipient,
 	out := &managementClient.Recipient{}
 
 	out.NotifierID = notifier.ID
-	if notifier.PagerdutyConfig != nil {
+	if notifier.DingtalkConfig != nil {
+		out.NotifierType = recipientTypeDingtalk
+	} else if notifier.MSTeamsConfig != nil {
+		out.NotifierType = recipientTypeMsTeams
+	} else if notifier.PagerdutyConfig != nil {
 		out.NotifierType = recipientTypePagerduty
 		out.Recipient = notifier.PagerdutyConfig.ServiceKey
 	} else if notifier.SlackConfig != nil {

--- a/rancher2/resource_rancher2_app_v2.go
+++ b/rancher2/resource_rancher2_app_v2.go
@@ -60,7 +60,7 @@ func resourceRancher2AppV2Create(d *schema.ResourceData, meta interface{}) error
 	if err != nil {
 		return fmt.Errorf("[ERROR] installing App V2: %s", err)
 	}
-	d.SetId(clusterID + appV2ClusterIDsep + chartInstallAction.Namespace + "/" + name)
+	d.SetId(clusterID + appV2ClusterIDsep + chartInstallAction.Namespace + "/" + d.Get("name").(string))
 
 	return resourceRancher2AppV2Read(d, meta)
 }

--- a/rancher2/schema_app_v2.go
+++ b/rancher2/schema_app_v2.go
@@ -65,6 +65,7 @@ func appV2Fields() map[string]*schema.Schema {
 		"chart_version": {
 			Type:        schema.TypeString,
 			Optional:    true,
+			Computed:    true,
 			Description: "Chart version",
 		},
 		"cluster_name": {

--- a/rancher2/schema_app_v2.go
+++ b/rancher2/schema_app_v2.go
@@ -147,7 +147,7 @@ func appV2Fields() map[string]*schema.Schema {
 		"wait": {
 			Type:        schema.TypeBool,
 			Optional:    true,
-			Default:     false,
+			Default:     true,
 			Description: "Wait until app is deployed",
 		},
 	}

--- a/rancher2/schema_recipient.go
+++ b/rancher2/schema_recipient.go
@@ -5,6 +5,8 @@ import (
 )
 
 const (
+	recipientTypeDingtalk  = "dingtalk"
+	recipientTypeMsTeams   = "msteams"
 	recipientTypePagerduty = "pagerduty"
 	recipientTypeSlack     = "slack"
 	recipientTypeSMTP      = "email"

--- a/rancher2/structure_app_v2.go
+++ b/rancher2/structure_app_v2.go
@@ -140,12 +140,17 @@ func expandChartInstallActionV2(in *schema.ResourceData, chartInfo *types.ChartI
 	if err != nil {
 		return nil, err
 	}
+	wait := in.Get("wait").(bool)
+	if len(chartIntalls) > 1 {
+		// Forcing wait = true if chart has dependencies
+		wait = true
+	}
 
 	timeOut := &metaV1.Duration{}
 	timeOut.Duration = in.Timeout(schema.TimeoutCreate)
 	obj := &types.ChartInstallAction{
 		Timeout:                  timeOut,
-		Wait:                     in.Get("wait").(bool),
+		Wait:                     wait,
 		DisableHooks:             in.Get("disable_hooks").(bool),
 		DisableOpenAPIValidation: in.Get("disable_open_api_validation").(bool),
 		Namespace:                namespace,
@@ -243,12 +248,17 @@ func expandChartUpgradeActionV2(in *schema.ResourceData, chartInfo *types.ChartI
 	if err != nil {
 		return nil, err
 	}
+	wait := in.Get("wait").(bool)
+	if len(chartUpgrades) > 1 {
+		// Forcing wait = true if chart has dependencies
+		wait = true
+	}
 
 	timeOut := &metaV1.Duration{}
 	timeOut.Duration = in.Timeout(schema.TimeoutUpdate)
 	obj := &types.ChartUpgradeAction{
 		Timeout:                  timeOut,
-		Wait:                     in.Get("wait").(bool),
+		Wait:                     wait,
 		DisableHooks:             in.Get("disable_hooks").(bool),
 		DisableOpenAPIValidation: in.Get("disable_open_api_validation").(bool),
 		Force:                    in.Get("force_upgrade").(bool),

--- a/rancher2/structure_app_v2.go
+++ b/rancher2/structure_app_v2.go
@@ -71,6 +71,27 @@ func expandChartInstallV2(in *schema.ResourceData, chartInfo *types.ChartInfo) (
 			"clusterName": in.Get("cluster_name").(string),
 		},
 	}
+	valuesData := v3.MapStringInterface{}
+	if v, ok := in.Get("values").(string); ok {
+		values, err := ghodssyamlToMapInterface(v)
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to unmarshal chart install values yaml: %#v", err)
+		}
+		if values == nil {
+			values = map[string]interface{}{}
+		}
+		if global, ok := values["global"].(map[string]interface{}); ok && len(global) > 0 {
+			if _, ok := global["cattle"].(map[string]interface{}); ok && len(global) > 0 {
+				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterId"] = globalInfo["cattle"].(map[string]interface{})["clusterId"]
+				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterName"] = globalInfo["cattle"].(map[string]interface{})["clusterName"]
+			} else {
+				values["global"].(map[string]interface{})["cattle"] = globalInfo["cattle"]
+			}
+		} else {
+			values["global"] = globalInfo
+		}
+		valuesData = v3.MapStringInterface(values)
+	}
 	if chartAnnotations, ok := chartInfo.Chart["annotations"].(map[string]interface{}); ok && len(chartAnnotations) > 0 {
 		if autoInstall, ok := chartAnnotations["catalog.cattle.io/auto-install"].(string); ok && len(autoInstall) > 0 {
 			chartAuto := splitBySep(autoInstall, "=")
@@ -87,41 +108,28 @@ func expandChartInstallV2(in *schema.ResourceData, chartInfo *types.ChartInfo) (
 				Version:     chartVersion,
 				ReleaseName: chartName,
 				Values: v3.MapStringInterface{
-					"global": globalInfo,
+					"global": valuesData["global"],
 				},
 			}
 			out = append(out, obj)
 		}
-		// Forcing release name and namespace if defined
-		if chartName, ok := chartAnnotations["catalog.cattle.io/release-name"].(string); ok && len(chartName) > 0 {
-			name = chartName
-		}
-		if chartNamespace, ok := chartAnnotations["catalog.cattle.io/namespace"].(string); ok && len(chartNamespace) > 0 {
-			namespace = chartNamespace
+		// Forcing release name and namespace if rancher certified
+		if chartCertified, ok := chartAnnotations["catalog.cattle.io/certified"].(string); ok && chartCertified == "rancher" {
+			if chartName, ok := chartAnnotations["catalog.cattle.io/release-name"].(string); ok && len(chartName) > 0 {
+				name = chartName
+				in.Set("name", name)
+			}
+			if chartNamespace, ok := chartAnnotations["catalog.cattle.io/namespace"].(string); ok && len(chartNamespace) > 0 {
+				namespace = chartNamespace
+				in.Set("namespace", namespace)
+			}
 		}
 	}
 	obj := types.ChartInstall{
 		ChartName:   chartInfo.Chart["name"].(string),
 		Version:     chartInfo.Chart["version"].(string),
 		ReleaseName: name,
-		Values:      v3.MapStringInterface{},
-	}
-	if v, ok := in.Get("values").(string); ok && len(v) > 0 {
-		values, err := ghodssyamlToMapInterface(v)
-		if err != nil {
-			return "", nil, fmt.Errorf("failed to unmarshal chart install values yaml: %#v", err)
-		}
-		if global, ok := values["global"].(map[string]interface{}); ok && len(global) > 0 {
-			if _, ok := global["cattle"].(map[string]interface{}); ok && len(global) > 0 {
-				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterId"] = globalInfo["cattle"].(map[string]interface{})["clusterId"]
-				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterName"] = globalInfo["cattle"].(map[string]interface{})["clusterName"]
-			} else {
-				values["global"].(map[string]interface{})["cattle"] = globalInfo["cattle"]
-			}
-		} else {
-			values["global"] = globalInfo
-		}
-		obj.Values = v3.MapStringInterface(values)
+		Values:      valuesData,
 	}
 	if v, ok := in.Get("annotations").(map[string]interface{}); ok && len(v) > 0 {
 		obj.Annotations = toMapString(v)
@@ -177,6 +185,27 @@ func expandChartUpgradeV2(in *schema.ResourceData, chartInfo *types.ChartInfo) (
 			"clusterName": in.Get("cluster_name").(string),
 		},
 	}
+	valuesData := v3.MapStringInterface{}
+	if v, ok := in.Get("values").(string); ok {
+		values, err := ghodssyamlToMapInterface(v)
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to unmarshal chart install values yaml: %#v", err)
+		}
+		if values == nil {
+			values = map[string]interface{}{}
+		}
+		if global, ok := values["global"].(map[string]interface{}); ok && len(global) > 0 {
+			if _, ok := global["cattle"].(map[string]interface{}); ok && len(global) > 0 {
+				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterId"] = globalInfo["cattle"].(map[string]interface{})["clusterId"]
+				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterName"] = globalInfo["cattle"].(map[string]interface{})["clusterName"]
+			} else {
+				values["global"].(map[string]interface{})["cattle"] = globalInfo["cattle"]
+			}
+		} else {
+			values["global"] = globalInfo
+		}
+		valuesData = v3.MapStringInterface(values)
+	}
 	if chartAnnotations, ok := chartInfo.Chart["annotations"].(map[string]interface{}); ok && len(chartAnnotations) > 0 {
 		if autoInstall, ok := chartAnnotations["catalog.cattle.io/auto-install"].(string); ok && len(autoInstall) > 0 {
 			chartAuto := splitBySep(autoInstall, "=")
@@ -193,43 +222,30 @@ func expandChartUpgradeV2(in *schema.ResourceData, chartInfo *types.ChartInfo) (
 				Version:     chartVersion,
 				ReleaseName: chartName,
 				Values: v3.MapStringInterface{
-					"global": globalInfo,
+					"global": valuesData["global"],
 				},
 				Force: in.Get("force_upgrade").(bool),
 			}
 			out = append(out, obj)
 		}
-		// Forcing release name and namespace if defined
-		if chartName, ok := chartAnnotations["catalog.cattle.io/release-name"].(string); ok && len(chartName) > 0 {
-			name = chartName
-		}
-		if chartNamespace, ok := chartAnnotations["catalog.cattle.io/namespace"].(string); ok && len(chartNamespace) > 0 {
-			namespace = chartNamespace
+		// Forcing release name and namespace if rancher certified
+		if chartCertified, ok := chartAnnotations["catalog.cattle.io/certified"].(string); ok && chartCertified == "rancher" {
+			if chartName, ok := chartAnnotations["catalog.cattle.io/release-name"].(string); ok && len(chartName) > 0 {
+				name = chartName
+				in.Set("name", name)
+			}
+			if chartNamespace, ok := chartAnnotations["catalog.cattle.io/namespace"].(string); ok && len(chartNamespace) > 0 {
+				namespace = chartNamespace
+				in.Set("namespace", namespace)
+			}
 		}
 	}
 	obj := types.ChartUpgrade{
 		ChartName:   chartName,
 		Version:     chartVersion,
 		ReleaseName: name,
-		Values:      map[string]interface{}{},
+		Values:      valuesData,
 		Force:       in.Get("force_upgrade").(bool),
-	}
-	if v, ok := in.Get("values").(string); ok && len(v) > 0 {
-		values, err := ghodssyamlToMapInterface(v)
-		if err != nil {
-			return "", nil, fmt.Errorf("failed to unmarshal chart upgrade values yaml: %#v", err)
-		}
-		if global, ok := values["global"].(map[string]interface{}); ok && len(global) > 0 {
-			if _, ok := global["cattle"].(map[string]interface{}); ok && len(global) > 0 {
-				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterId"] = globalInfo["cattle"].(map[string]interface{})["clusterId"]
-				values["global"].(map[string]interface{})["cattle"].(map[string]interface{})["clusterName"] = globalInfo["cattle"].(map[string]interface{})["clusterName"]
-			} else {
-				values["global"].(map[string]interface{})["cattle"] = globalInfo["cattle"]
-			}
-		} else {
-			values["global"] = globalInfo
-		}
-		obj.Values = v3.MapStringInterface(values)
 	}
 	if v, ok := in.Get("annotations").(map[string]interface{}); ok && len(v) > 0 {
 		obj.Annotations = toMapString(v)

